ledger/src/tests.rs

```diff
@@ -2668,4 +2669,141 @@ mod valid_solutions {
+        let crate::test_helpers::TestEnv { ledger, private_key, address, .. } =
+            crate::test_helpers::sample_test_env(rng);
+
+        // Retrieve the puzzle parameters.
+        let puzzle = ledger.puzzle();
+        let latest_epoch_hash = ledger.latest_epoch_hash().unwrap();
+        let minimum_proof_target = ledger.latest_proof_target();
+
+        // Initialize a valid solution object.
+        let mut valid_solution = None;
+        while valid_solution.is_none() {
+            let solution = puzzle.prove(latest_epoch_hash, address, rng.gen(), None).unwrap();
+            if puzzle.get_proof_target(&solution).unwrap() >= minimum_proof_target {
+                valid_solution = Some(solution);
+            }
+        }
+        // Unwrap the valid solution.
+        let valid_solution = valid_solution.unwrap();
+
+        // Construct a malicious solution with a different target.
+        let malicious_solution =
+            Solution::new(*valid_solution.partial_solution(), valid_solution.target().saturating_add(1));
+
+        assert_eq!(
+            valid_solution.id(),
+            malicious_solution.id(),
+            "The malicious solution should have the same ID as the valid solution"
+        );
+
+        // Create a valid transaction for the block.
+        let inputs = [Value::from_str(&format!("{address}")).unwrap(), Value::from_str("10u64").unwrap()];
+        let transfer_transaction = ledger
+            .vm
+            .execute(&private_key, ("credits.aleo", "transfer_public"), inputs.iter(), None, 0, None, rng)
+            .unwrap();
+
+        // Check that the block creation fixes the malformed solution.
+        let expected_solution_id = valid_solution.id();
+        let expected_solution = valid_solution;
+        let mut check_block = |candidate_solutions: Vec<Solution<CurrentNetwork>>| {
+            // Create a block.
+            let block = ledger
+                .prepare_advance_to_next_beacon_block(
+                    &private_key,
+                    vec![],
+                    candidate_solutions.clone(),
+                    vec![transfer_transaction.clone()],
+                    rng,
+                )
+                .unwrap();
+
+            // Check that the next block is valid.
+            ledger.check_next_block(&block, rng).unwrap();
+
+            // Check that the block's solutions are well-formed.
+            assert_eq!(block.solutions().len(), 1);
+            assert_eq!(block.aborted_solution_ids().len(), 0);
+
+            // Fetch the solution from the block.
+            let (solution_id, solution) = block.solutions().as_ref().unwrap().first().unwrap();
+            assert_eq!(*solution_id, expected_solution_id, "Check that the block has the correct solution ID");
+            assert_eq!(*solution, expected_solution, "Check that the block has the correct solution");
+        };
+
+        // Case 1: The valid solution is included in the block.
```

Comment:
For completeness, could you also try reversing the cases with a second solution (first inserting the malicious solution, then the valid one).

Reply:
You shouldn't be able to have both in the block, since snarkOS filters out duplicates based on the `solutionID`. 

We are just trying to prevent a deviation case where different nodes processes the same solution ID with different solutions.

Reply:
The `ledger` object is re-used across tests, potentially doing caching of the solution target. So I think a slightly more proper test would be to let Case 1 be malicious, and Case 2 succesful.

Reply:
Resolved with [8f2c014](https://github.com/AleoNet/snarkVM/pull/2520/commits/8f2c0144c77a164e066f253153001dd072ec6103)